### PR TITLE
Optimize while fetching authorized coverages for an user

### PR DESCRIPTION
**Pull request by kadhikari**
Fetching list of coverage for a request without coverage:  
**Before:** 
* Fetch all the coverages where **from** and **to** objects could be found. This function requires verifying existence of coord/pt_object in each coverage (all coverages) even if **cached**.
* Filter authorized coverages for the user among coverages found above.  

**After:**
* For an user, get list of authorized coverages (free + private)
* Filter coverages for coord/objects in the request without coverage

Note: The work is not finished yet. There are some test fails.

### CODE REVIEW

File: source/jormungandr/jormungandr/instance_manager.py [hunk 1]
```diff
@@ -219,6 +219,25 @@ def stop(self):
         if not self.thread_event.is_set():
             self.thread_event.set()
 
+    def _get_free_instances(self):
+        instances = []
+        for name, instance in self.instances.items():
+            instance_db = instance.get_models()
+            if instance_db.is_free:
+                instances.append(instance)
```
Comment (azime) [inline source/jormungandr/jormungandr/instance_manager.py]: use [property](https://github.com/hove-io/navitia/blob/dev/source/jormungandr/jormungandr/instance.py#L491) of instance:
`
if instance.is_free():
    instances.append(instance)
`

File: source/jormungandr/jormungandr/instance_manager.py [hunk 2]
```diff
@@ -261,6 +295,30 @@ def _all_keys_of_id(self, object_id):
             raise RegionNotFound(object_id=object_id)
         return instances
 
+    def _all_keys_of_id_in_instances(self, instances, object_id):
+        valid_instances = []
+        for instance in instances:
+            if self._exists_id_in_instance(instance, object_id):
+                valid_instances.append(instance)
+        if not instances:
+            raise RegionNotFound(object_id=object_id)
```
Comment (azime) [inline source/jormungandr/jormungandr/instance_manager.py]: if not instances:
            raise RegionNotFound(object_id=object_id)
        valid_instances = []
        for instance in instances:
            if self._exists_id_in_instance(instance, object_id):
                valid_instances.append(instance)

File: source/jormungandr/jormungandr/instance_manager.py (hunk 1 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/instance_manager.py]: :+1:

File: source/jormungandr/jormungandr/instance_manager.py (hunk 2 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/instance_manager.py]: :+1:

File: source/jormungandr/jormungandr/instance_manager.py [hunk 3]
```diff
@@ -219,6 +219,18 @@ def stop(self):
...
+            for name, i in self.instances.items()
+            if authentication.has_access(name, abort=False, user=user, api=api)
+        ]
+
+        if not authorized_instances:
+            context = 'User has no access to any instance'
+            authentication.abort_request(user, context)
+        return authorized_instances
+
     def _filter_authorized_instances(self, instances, api):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/instance_manager.py]: so this function is no longer used, right? can it be removed?

**Review by xlqian — APPROVED** [review]

**Review by azime — APPROVED** [review]

File: source/jormungandr/tests/authentication_tests.py [hunk 4]
```diff
@@ -262,15 +262,15 @@ def test_status_code(self):
...
         """
-        the authentication process must not mess if the region is not found
+        the authentication process prevails even if the region is not found
         """
         with user_set(app, FakeUserAuth, 'bob'):
             r, status = self.query_no_assert('/v1/coverage/the_marvelous_unknown_region/stop_areas')
 
-            assert status == 404
-            assert 'error' in r
+            assert status == 403
```
Comment (xlqian) [inline source/jormungandr/tests/authentication_tests.py]: this should be 404, I'm working on it

Comment (kadhikari) [reply source/jormungandr/tests/authentication_tests.py]: :+1:

File: source/jormungandr/jormungandr/instance_manager.py [hunk 5]
```diff
@@ -284,27 +296,34 @@ def get_regions(self, region_str=None, lon=None, lat=None, object_id=None, api='
             return [i.name for i in valid_instances]
 
     def get_instances(self, name=None, lon=None, lat=None, object_id=None, api='ALL'):
-        available_instances = []
+        if name and name not in self.instances:
```
Comment (xlqian) [inline source/jormungandr/jormungandr/instance_manager.py:285]: @kadhikari  I added this line to handle the 404 cases

Comment (kadhikari) [reply source/jormungandr/jormungandr/instance_manager.py:285]: :+1:

File: source/jormungandr/jormungandr/instance_manager.py (hunk 3 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/instance_manager.py]: :+1:

**Review by kadhikari — COMMENTED** [review]
Modifications by @xlqian are fine to me
Thanks

**Review by pbench — APPROVED** [review]

**Pull request merged**